txnsync/service_test.go

```diff
@@ -0,0 +1,138 @@
+package txnsync
+
+import (
+	"encoding/binary"
+	"math/rand"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/logging"
+	"github.com/algorand/go-algorand/util/execpool"
+	"github.com/algorand/go-algorand/util/timers"
+)
+
+type mockLogger struct {
+	logging.Logger
+}
+
+type mockNodeConnector struct {
+	NodeConnector
+	calledEvents *bool
+}
+
+func makeMockNodeConnector(calledEvents *bool) mockNodeConnector {
+	return mockNodeConnector{calledEvents: calledEvents}
+}
+
+func (fn *mockNodeConnector) Events() <-chan Event {
+	*fn.calledEvents = true
+	return nil
+}
+
+func (fn *mockNodeConnector) GetCurrentRoundSettings() (out RoundSettings) { return }
+
+func (fn *mockNodeConnector) Clock() (out timers.WallClock) {
+	return timers.MakeMonotonicClock(time.Now())
+}
+
+func (fn *mockNodeConnector) Random(rng uint64) uint64 {
+	var xb [8]byte
+	rand.Read(xb[:])
+	rv := binary.LittleEndian.Uint64(xb[:])
+	return rv % rng
+}
+
+func (fn *mockNodeConnector) GetPeers() []PeerInfo { return nil }
+
+func (fn *mockNodeConnector) GetPeer(interface{}) (out PeerInfo) { return }
+
+func (fn *mockNodeConnector) UpdatePeers(txsyncPeers []*Peer, netPeers []interface{}, peersAverageDataExchangeRate uint64) {
+}
+func (fn *mockNodeConnector) SendPeerMessage(netPeer interface{}, msg []byte, callback SendMessageCallback) {
+}
+func (fn *mockNodeConnector) GetPendingTransactionGroups() (txGroups []transactions.SignedTxGroup, latestLocallyOriginatedGroupCounter uint64) {
+	return
+}
+func (fn *mockNodeConnector) IncomingTransactionGroups(peer *Peer, messageSeq uint64, txGroups []transactions.SignedTxGroup) (transactionPoolSize int) {
+	return 0
+}
+func (fn *mockNodeConnector) NotifyMonitor() chan struct{} { return nil }
+
+type mockThreadPool struct {
+	execpool.BacklogPool
+}
+
+// TestStartStopTransactionSyncService test that we can start and stop the transaction sync service
+func TestStartStopTransactionSyncService(t *testing.T) {
+
+	calledEventsInNodeConnector := false
+
+	a := require.New(t)
+
+	mLogger := mockLogger{}
+	mNodeConnector := makeMockNodeConnector(&calledEventsInNodeConnector)
+	cfg := config.Local{}
+	mThreadPool := mockThreadPool{}
+
+	hashDigest := crypto.Hash([]byte{0x41, 0x6b, 0x69, 0x6b, 0x69})
+
+	service := MakeTransactionSyncService(mLogger, &mNodeConnector, true, "GENID", hashDigest, cfg, mThreadPool)
+
+	a.NotNil(service)
+
+	service.Start()
+	service.Stop()
+
+	a.True(calledEventsInNodeConnector)
```

Comment:
I think that this test might randomally fail and could result in a data race.

Reply:
Isn't the Events() function in mainloop.go:124 called before any of the channel switch statements?  The Stop() function also waits for the goroutine to run right?